lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+	int index_next = INVALID;
+
+	/* Find the next available list index. */
+	for (int i = 0; i < ARRAY_SIZE(ctx.list_internal) - 1; i++) {
+		if (ctx.list_internal[i].message.notified_count == 0) {
+			index_next = i;
+			break;
+		}
+	}
+
+	if (index_next == INVALID) {
+		LOG_ERR("No available entries in pending message list");
+		return -ENOMEM;
+	}
+
+	ctx.list_internal[index_next].message = *message;
+
+	sys_slist_append(&ctx.pending_list, &ctx.list_internal[index_next].header);
+	return index_next;
+}
+
+static int list_remove(uint32_t id)
+{
+	struct qos_metadata *node = NULL, *next_node = NULL, *node_prev = NULL;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST,
+	};
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&ctx.pending_list, node, next_node, header) {
+		if (node->message.id == id) {
+			evt.message = node->message;
+			notify_event(&evt);
+			memset(&node->message, 0, sizeof(struct qos_data));
+			sys_slist_remove(&ctx.pending_list, &node_prev->header, &node->header);
+			return 0;
+		}
+
+		node_prev = node;
+	};
+
+	return -ENODATA;
+}
+
+/* Public API functions */
+
+int qos_init(qos_evt_handler_t evt_handler)
+{
+	if (evt_handler == NULL) {
+		LOG_DBG("Handler is NULL");
+		return -EINVAL;
+	}
+
+	if (ctx.initialized) {
+		return -EALREADY;
+	}
+
+	ctx.initialized = true;
+
+	LOG_DBG("Registering handler %p", evt_handler);
+	ctx.app_evt_handler = evt_handler;
+
+	/* Initializing linked list and delayed work. */
+	sys_slist_init(&ctx.pending_list);
+	k_work_init_delayable(&ctx.timeout_handler_work, timeout_handler_work_fn);
+	return 0;
+}
+
+int qos_message_add(struct qos_data *message)
+{
+	int err = 0, ret;
+	struct qos_evt evt = {
+		.type = QOS_EVT_MESSAGE_NEW,
+		.message = *message
+	};
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	if (message == NULL) {
+		LOG_WRN("Message cannot be NULL");
+		err = -EINVAL;
+		goto exit;
+	}
+
+	/* Only ACK_REQUIRED messages are added to the internal list. */
+	if (qos_message_has_flag(message, QOS_FLAG_RELIABILITY_ACK_REQUIRED)) {
+		ret = list_append(message);
+		if (ret < 0) {
+			LOG_WRN("No list entries available, error: %d", ret);
+			evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+			notify_event(&evt);
+			err = -ENOMEM;
+			goto exit;
+		}
+
+		ctx.list_internal[ret].message.notified_count++;
+		notify_event(&evt);
+	} else {
+		notify_event(&evt);
+
+		/* If the message does not carry the ACK_REQUIRED flag, its removed from the
+		 * internal list immediately.
+		 */
+		evt.type = QOS_EVT_MESSAGE_REMOVED_FROM_LIST;
+		notify_event(&evt);
+	}
+
+	timer_start();
+
+exit:
+	k_mutex_unlock(&ctx_lock);
+	return err;
+}
+
+int qos_message_remove(uint32_t id)
+{
+	int err;
+
+	k_mutex_lock(&ctx_lock, K_FOREVER);
+
+	err = list_remove(id);
+	if (err) {
+		goto exit;
+	}
+
+	/* If the removed message is the last in the pending list, we stop the internal timer
+	 * and clear the retry counter.
+	 */
+	if (sys_slist_is_empty(&ctx.pending_list)) {
+		LOG_WRN("QoS list is empty!");
```

Comment:
Isn't this part of normal operation? Should this be `LOG_DBG` instead?

Reply:
Yes. Debug residue.